### PR TITLE
 preview plot fit

### LINKED ISSUE #15304: MantidWidget for plotting results of a Curve Fitting
A widget containing two PreviewPlot widgets, one for comparison of data against fit (as well as a guess), and a second PreviewPlot for the residuals.

Requirements:
1. The range selectors (blue lines) should be tied in both plots. Modifying one causes modification of the other.
2. The curves to plot are data, guess, fit, and residuals. Only residuals goes to the below plot.
3. The widget presents handles (or pointers) to all curves in the same way, irrespective of which PreviewPlot panel is used for their representation.

### PR DESCRIPTION
**Pull request by jmborr**
Fixes #15304

No release notes

Custom interface DisplayCurveFitTest has been written with the sole purpose to test new MantidWidget DisplayCurveFit.
1. Download file [fit_results_example.txt](https://github.com/mantidproject/mantid/files/139371/fit_results_example.txt). This file, when loaded, will create a workspace with four spectra containing data for the data-curve, fit-curve, residuals-curve, and guess-curve.
2. Compile in Debug mode, otherwise the interface will be unavailable.
3. Open MantidPlot, then Interfaces -> DynamicPDF -> Test DisplayCurveFit widget.
4. Load file fit_results_example.txt in the interface. The results should look something like this:
   displaycurvefittest
5. Deleting workspace <code>fit_results_example</code> will delete the curves too.
6. Try to move the two range selectors (vertical dashed lines). Checking the "update range selectors?" and reloading file <i>fit_results_example.txt</i> will reset the position of the range selectors to the original positions.

Also, please verify that the plugin for the DisplayCurveFit widget is availabe in Qt-Designer. Follow instructions in the [appropriate Mantid Wikipage](http://www.mantidproject.org/Qt_designer_with_custom_plugins).

### CODE REVIEW

**Comment by FedeMPouzols** [comment]
@shahrozahmed1 : I'll be testing this on linux. Can you give it a look on windows? This is a good example to look at in order to get familiar with the fit-plot widgets.

**Comment by shahrozahmed1** [comment]
While testing on Windows, I came across an issues while following step 5 provided above. When trying to delete workspace `fit_results_example`, it produces an error instead of just deleting the curves.
capture

Also if the file loaded to this GUI will always be a text file then another feature which could be included is, while browsing a file, enable it to search for .txt file instead of Data Files.

**Comment by jmborr** [comment]
@shahrozahmed1 This is just a GUI for purposes of testing, so the format of the file to be loaded is not important. Also, can you please build in <b>release</b> mode and see if you get this error when deleting the workspace?

**Comment by shahrozahmed1** [comment]
@jmborr @FedeMPouzols After building twice on my machine `Test DisplayCurveFit widget` interface could be found in the `DynamicPDF` category on Windows **Release** mode.

**Comment by FedeMPouzols** [comment]
I think that's the intended behavior, right? If you compile it in release mode it is not registered:
https://github.com/mantidproject/mantid/pull/15391/files#diff-e58764e0ac279b10c29f1570105eea91R20

**Comment by jmborr** [comment]
Arrgggg I apologize!!!! I had completely forgotten about the ifndef. I am going to remove this condition so that shahrozahmed1 can build in release mode. I suspect that the ASSERT failure that shahrozahmed1 found is evaluated only when in Debug mode, so I wonder how the testing interface will perform if compiled in Release model.

File: MantidQt/CustomInterfaces/src/DynamicPDF/DisplayCurveFitTest.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+ * @brief The test proper that loads the fit curves to be
+ * displayed and the two ranges.
+ * @param workspaceName the name of the workspace containing
+ * the data of the curves to be displayed.
+ */
+void DisplayCurveFitTest::loadSpectra(const QString &workspaceName) {
+  auto workspace = Mantid::API::AnalysisDataService::Instance()
+                       .retrieveWS<Mantid::API::MatrixWorkspace>(
+                           workspaceName.toStdString());
+  if (workspace->getNumberHistograms() < 4) {
```
Comment (FedeMPouzols) [inline MantidQt/CustomInterfaces/src/DynamicPDF/DisplayCurveFitTest.cpp]: Here `workspace` is not checked to be `!= nullptr` before being used.
I can imagine that coverity is going to complain about this "potential null dereference".

**Comment by FedeMPouzols** [comment]
This works well for me on debian, and it makes sense if it is only available in Debug mode. Maybe "Utility" would be a better place for it, as it is not specific to PDF or any other technique?

**Comment by FedeMPouzols** [comment]
Just notice one small detail. Was "Dynamic PDF Slice Selector" the intended name for this window?

**Comment by jmborr** [comment]
@FedeMPouzols I'll add a check for null workspace pointer. I had not done it because loadSpectra() is solely used as a SLOT for a SIGNAL emited by the widget DataSelector, which is only emitted if the workspace is successfully loaded from a file or if present in the AnalysisDataService. However, it had not occurred to me that the user may try to load or use a workspace of different type than a matrix workspace.

**Comment by jmborr** [comment]
@FedeMPouzols I plan to morph this interface into the "background remover" interface after I include a custom FitPropertyBrowser widget. I could have written everything together but I favor doing one step at a time.

**Comment by FedeMPouzols** [comment]
@jmborr : I've just written a last comment about an exception. I think this is ready to go. I'm transferring the test-ownership to @shahrozahmed1 . @jmborr let @shahrozahmed1 know if this needs any more testing on windows. Otherwise, just add the "ship it" emoji/tag so that a gatekeeper merges this in.

**Comment by jmborr** [comment]
I have substituted the "throw" statements with warning popups.

To test the popup asserting the wrong type of workspace:
* Use algorithm "CreateEmptyTableWorkspace"
* Try to load the created table workspace

To test the popup asserting an unsufficient number of spectra:
* Load  fit_results_example.txt
* Use algorithm "ExtractSingleSpectrum" to create a new workspace with only one spectrum
* Try to load the created workspace

**Comment by shahrozahmed1** [comment]
the two tests listed above worked fine on Windows :+1: 

-- :shipit: --

**Comment by jmborr** [comment]
@shahrozahmed1 Can you close the pull request, then?

**Comment by FedeMPouzols** [comment]
@jmborr : we now have this "Second Reviewer (gatekeepers)" model for merging PRs. (see the first comment in this PR). As I was involved in the first testing, I shouldn't merge it myself, but another gatekeeper will take a look soon.

**Comment by jmborr** [comment]
Oh, yes, I had forgotten about it. Thanks!

**Comment by jmborr** [comment]
@quantumsteve Thanks!

**Pull request merged**